crates/iota-indexer/src/apis/governance_api.rs

```diff
@@ -561,6 +563,50 @@ impl GovernanceReadApi {
     }
 }
 
+/// Backfill missing rates for some epochs due to safe mode. If a rate is
+/// missing for epoch e, we will use the rate for epoch e-1 to fill it. Rates
+/// returned are in descending order by epoch.
+fn backfill_rates(
+    mut rates: Vec<(EpochId, PoolTokenExchangeRate)>,
+) -> Vec<(EpochId, PoolTokenExchangeRate)> {
+    if rates.is_empty() {
+        return rates;
+    }
+    // ensure epochs are processed in increasing order
+    rates.sort_unstable_by_key(|(epoch_id, _)| *epoch_id);
+
+    // Check if there are any gaps in the epochs
+    let (min_epoch, _) = rates.first().expect("rates should not be empty");
+    let (max_epoch, _) = rates.last().expect("rates should not be empty");
+    let expected_len = (max_epoch - min_epoch + 1) as usize;
+
+    // Only perform backfilling if there are gaps
+    if rates.len() == expected_len {
+        rates.reverse();
+        return rates;
+    }
+
+    let mut filled_rates: Vec<(EpochId, PoolTokenExchangeRate)> = Vec::with_capacity(expected_len);
+    let mut missing_rates = Vec::new();
```

Comment:
We already know how many rates are missing at this point. We could initialise with a capacity already.

Reply:
Thanks. Missed this. Added at [86467c0](https://github.com/iotaledger/iota/pull/5184/commits/86467c034510d51e2d2423220ff7e4d2000c9839).